pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/MeasuresFunctors.hpp

```diff
@@ -16,27 +16,426 @@
 #include <Kokkos_Core.hpp>
 #include <Kokkos_StdAlgorithms.hpp>
 
+#include "UtilKokkos.hpp"
+
 namespace Pennylane::LightningKokkos::Functors {
-/**
- * @brief Compute probability distribution from StateVector.
- *
- * @param arr_ StateVector data.
- * @param probabilities_ Discrete probability distribution.
- */
-template <class PrecisionT> struct getProbFunctor {
-    Kokkos::View<Kokkos::complex<PrecisionT> *> arr;
-    Kokkos::View<PrecisionT *> probability;
 
-    getProbFunctor(Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
-                   Kokkos::View<PrecisionT *> probability_)
-        : arr(arr_), probability(probability_) {}
+template <class PrecisionT, class DeviceType> class getProbsFunctor {
+  public:
+    // Required for functor:
+    using execution_space = DeviceType;
+    using value_type = PrecisionT[];
+    const unsigned value_count;
+
+    using ComplexT = Kokkos::complex<PrecisionT>;
+    Kokkos::View<ComplexT *> arr;
+    Kokkos::View<std::size_t *> all_indices;
+    Kokkos::View<std::size_t *> all_offsets;
+    getProbsFunctor(const Kokkos::View<ComplexT *> &arr_,
+                    [[maybe_unused]] const std::vector<std::size_t> &wires_,
+                    const Kokkos::View<std::size_t *> all_indices_,
+                    const Kokkos::View<std::size_t *> all_offsets_)
+        : value_count{1U << wires_.size()}, arr{arr_},
+          all_indices{all_indices_}, all_offsets{all_offsets_} {}
 
     KOKKOS_INLINE_FUNCTION
-    void operator()(const std::size_t k) const {
-        const PrecisionT REAL = arr(k).real();
-        const PrecisionT IMAG = arr(k).imag();
-        probability(k) = REAL * REAL + IMAG * IMAG;
+    void init(PrecisionT dst[]) const {
+        for (unsigned i = 0; i < value_count; ++i)
+            dst[i] = 0.0;
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    void join(PrecisionT dst[], const PrecisionT src[]) const {
+        for (unsigned i = 0; i < value_count; ++i)
+            dst[i] += src[i];
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    void operator()(const std::size_t i, const std::size_t j,
+                    PrecisionT dst[]) const {
+        const std::size_t index = all_indices(i) + all_offsets(j);
+        const PrecisionT rsv = arr(index).real();
+        const PrecisionT isv = arr(index).imag();
+        dst[i] += rsv * rsv + isv * isv;
+    }
+};
+
+template <class PrecisionT, class DeviceType, std::size_t num_wires>
+class getProbsNQubitOpFunctor {
```

Comment:
This is used with the Range policy (parallelized over offsets which are computed on-the-fly), good performance when #qubits >> #targets; especially when the template parameter `num_qubits != 0`).